### PR TITLE
fix nf4 performance in hybrid CPU

### PR DESCRIPTION
**Pull request by yuchengliu1**
## Type of Change

Automatic modify the thread number and disable to run nf4 on E-core when inferencing nf4 model on hybrid CPU.
Quantizing to nf4 model on hybrid CPU will get a warning now.

## Description

detail description 
Issues: xxx

## Expected Behavior & Potential Risk

the expected behavior that triggered by this PR 

## How has this PR been tested?

how to reproduce the test (including hardware information)

## Dependency Change?

any library dependency introduced or removed

### CODE REVIEW

File: neural_speed/models/baichuan/baichuan.cpp [hunk 1]
```diff
@@ -106,6 +106,13 @@ static bool baichuan_model_eval_internal(model_context* ctx, const model_input*
   // otherwise, the threads are spin-lock waiting for the BLAS calls and are degrading the performance
   ne_cgraph gf = {};
   gf.n_threads = N >= 32 && ne_cpu_has_blas() ? 1 : n_threads;
+  if (hparams.ftype == NE_FTYPE_MOSTLY_NF4 && bestla_is_hybrid() && gf.n_threads > bestla_get_Pcore_number()) {
+    printf(
+        "WARNING: NF4 is poor at Ecore, only use P-core to inference. Not use the thread number according to settings. "
+        "New thread number is: %d\n.",
+        bestla_get_Pcore_number());
+    gf.n_threads = bestla_get_Pcore_number();
+  }
```
Comment (airMeng) [inline neural_speed/models/baichuan/baichuan.cpp:115]: I will prefer to subtract out these logic so you can keep improving within bestla

```cpp
// ne_bestla.cpp
int bestla_get_best_thread_num(ne_ftype ft){
  if (ftype == NE_FTYPE_MOSTLY_NF4 && bestla_is_hybrid()) return bestla_get_Pcore_number();
  else return ${max_thread_num}
}

// various models.cpp
int best_n_thread = bestla_get_best_thread_num(hparams.ftype);
 gf.n_threads = N >= 32 && ne_cpu_has_blas() ? 1 : (best_n_thread < n_threads : best_n_threads : n_threads);

**Review by zhewang1-intc — COMMENTED** [review]
pls note that the perf of fp4_e2m1 & fp4_bnb are also poor on E-core.

File: neural_speed/models/baichuan/baichuan.cpp (hunk 1 above)
Comment (yuchengliu1) [reply neural_speed/models/baichuan/baichuan.cpp]: Done

**Comment by kevinintel** [comment]
Please remember to use int4 as default when hybrid

File: neural_speed/models/model_utils/util.cpp [hunk 2]
```diff
@@ -47,3 +49,21 @@ int32_t get_num_physical_cores() {
...
+    else
+      res = bestla_get_thread_number();
+  }
+  if (res < n_threads) {
+    printf("WARNING: Thread number exceed the limit. Actual thread number is: %d\n now.", res);
+  } else {
+    res = n_threads;
+  }
+  return res;
+}
```
Comment (zhewang1-intc) [inline neural_speed/models/model_utils/util.cpp:69]: why cpplint not thrown an warning?

Comment (DDEle) [reply neural_speed/models/model_utils/util.cpp:69]: Because we are using clang-tidy rather than cpplint now?

**Comment by luoyu-intel** [comment]
@yuchengliu1 Can you check whether the new thread pool dispatch all jobs to P cores if E cores have poor performance?

**Comment by luoyu-intel** [comment]
f4 performance will be remeasured after #172.

**Comment by yuchengliu1** [comment]
> @yuchengliu1 Can you check whether the new thread pool dispatch all jobs to P cores if E cores have poor performance?

Some parallel created with schedule2D do not dispatch, such as parallel of PrologueA and MHA reorder

**Comment by luoyu-intel** [comment]
> > @yuchengliu1 Can you check whether the new thread pool dispatch all jobs to P cores if E cores have poor performance?
> 
> Some parallel created with schedule2D do not dispatch, such as parallel of PrologueA and MHA reorder

@yuchengliu1 Just run benchmark. Focus on GEMM only.

**Pull request closed without merging**